ethcore/verification/src/queue/mod.rs

```diff
@@ -590,6 +591,22 @@ impl<K: Kind, C> VerificationQueue<K, C> {
 			&& v.verified.load_len() == 0
 	}
 
+	/// Returns true, if in processing queue there is no descendant of the current best block
+	/// May return false negative for longer queues
+	pub fn processing_fork(&self, best_block_hash: &H256) -> bool {
+		let processing = self.processing.read();
+		if processing.is_empty() || processing.len() > MAX_QUEUE_WITH_FORK {
```

Comment:
So this optimization is introduced to avoid spending too much time traversing the entire `processing` map, right?
If we introduce a different mapping of `ParentHash -> count_of_blocks_in_processing_that_build_on_top` we could do it in constant time for any parent hash.

Reply:
I implemented this approach, tested it and found one drawback: it doesn't work well in the case, when node is catching up blocks. In this case imported blocks are inserted into the queue one by one, as a result the suggested data structure contains N parent hashes with count 1. N could be pretty huge. So I considered it as a significant overhead. As I didn't find good solution to address both situations, I've decided to revert it and leave the current approach for now.